### ISSUE TITLE
Document behaviour of mixed derivatives

### ISSUE THREAD

**Issue by Jesperoka** [google/jax#15082]
It might exist, but I couldn't find any examples on how ordering of derivatives act, for scalar functions with multiple vector inputs. It would be nice if this could exist in [The Autodiff Cookbook](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html) or [Advanced Automatic Differentiation in JAX](https://jax.readthedocs.io/en/latest/jax-101/04-advanced-autodiff.html).

Here's an example to explain what I mean:
```python
from jax import jacfwd, grad
from jax import numpy as jnp

def M1(x, y):
    return x[0]*y[0]*jnp.array(\
        [[1, 2],
         [3, 4]], dtype=jnp.float64)

def M2(x):
    return x[0]*x[0]*jnp.array(\
        [[0.1, 0.2],
         [0.3, 0.4]], dtype=jnp.float64)

def f(x: jnp.ndarray, y: jnp.ndarray):
    return jnp.squeeze(jnp.transpose(x) @ M1(x, y) @ y + jnp.transpose(y) @ y - jnp.transpose(y) @ M2(x) @ x)

x = jnp.array([[4],[2]], dtype=jnp.float64) # full dimensional inputs
y = jnp.array([[6],[9]], dtype=jnp.float64) # full dimensional  inputs
xSqzd = jnp.squeeze(jnp.array([[4],[2]], dtype=jnp.float64)) # squeezed inputs
ySqzd = jnp.squeeze(jnp.array([[6],[9]], dtype=jnp.float64)) # squeezed inputs

gradX = grad(f, 0)
jacX = jacfwd(f, 0)
gradY = grad(f, 1)
jacY = jacfwd(f, 1)

# Printing
def printTheStuff(gradX, jacX, gradY, jacY, x, y):
    print("gradX\n", gradX(x, y), "shape:", gradX(x, y).shape)
    print("jacX\n", jacX(x, y), "shape:", jacX(x, y).shape)
    print("gradY\n", gradY(x, y), "shape:", gradY(x, y).shape)
    print("jacY\n", jacY(x, y), "shape:", jacY(x, y).shape)
    print("\njacfwd(gradX, 1)\n", jacfwd(gradX, 1)(x, y), "shape:", jacfwd(gradX, 1)(x, y).shape)
    print("\njacfwd(jacX, 1)\n", jacfwd(jacX, 1)(x, y), "shape:", jacfwd(jacX, 1)(x, y).shape)
    print("\njacfwd(gradY, 0)\n", jacfwd(gradY, 0)(x, y), "shape:", jacfwd(gradY, 0)(x, y).shape)
    print("\njacfwd(jacY, 0)\n", jacfwd(jacY, 0)(x, y), "shape:", jacfwd(jacY, 0)(x, y).shape)
    
# Without squeezed inputs
print("------- No squeeze ---------------------------------")
printTheStuff(gradX, jacX, gradY, jacY, x, y)
print("----------------------------------------------------")

print("\n\n------- Reshaped (2, 2) ----------------------------")
print("jnp.reshape(jacfwd(gradX, 1), (2,2))\n", jnp.reshape(jacfwd(gradX, 1)(x, y), (2,2)))
print("\njnp.reshape(jacfwd(jacX, 1), (2,2))\n", jnp.reshape(jacfwd(jacX, 1)(x, y), (2,2)))
print("\njnp.reshape(jacfwd(gradY, 0), (2,2))\n", jnp.reshape(jacfwd(gradY, 0)(x, y), (2,2)))
print("\njnp.reshape(jacfwd(jacY, 0), (2,2))\n", jnp.reshape(jacfwd(jacY, 0)(x, y), (2,2)))
print("----------------------------------------------------")

# With  squeezed inputs
print("\n\n------- Squeeze ------------------------------------")
printTheStuff(gradX, jacX, gradY, jacY, xSqzd, ySqzd)
print("----------------------------------------------------")
```

The output of the prints is then: 

![jaxOutput](https://user-images.githubusercontent.com/79093095/226179036-4b3f38c8-7d80-4784-b9b0-32c75d9c8a73.png)

Where we can see that the ordering of the derivatives taken has the effect of transposing the result. 

Since jacobians, gradients, and total derivatives (and their notation) is already easy to make mistakes with, it would be nice if the documentation could include examples of this kind of thing (since it already includes the normal hessian case), with the correponding math given alongside.

Also, although it doesn't matter here, it's notable that the jax jacobian is *not* the transpose of the jax gradient of the same (scalar) function, since that's a quite common convention.

I'm sure other linear-algebraically challenged individuals out there would also appreciate this.

And if anyone else is often confused by why order matters:
![Page 1](https://user-images.githubusercontent.com/79093095/226185224-620c71ac-72c4-496b-b4b6-a94184cb86e7.PNG)

**Comment by jakevdp** [comment]
Hi, thanks for the question. It looks like you explored this pretty deeply!

Do you have an idea of specifically what you would say and where in the documentation you would say it to clarify this?

For the record, I don't think this is fundamentally an issue of how JAX treats mixed derivatives, but rather how JAX transformations compose. If you have a function $f: \mathbb{R}^N \to \mathbb{R}^M$, the signature of the JVP will have signature $f^\prime: \mathbb{R}^N \to \mathbb{R}^{(N, M)}$. I think everything else about the shape of mixed derivatives follows logically from that. And that is true whatever `f` represents: a simple function, a derivative w.r.t. its first argument, a derivative w.r.t its second argument, etc.

**Comment by Jesperoka** [comment]
Yes, you are right about the composition of JAX transformations of course, and I agree that everything follows logically. For some context, I was at first confused not because of anything to do with JAX, but because I found the math notation $\frac{\partial^2{\cal L}}{\partial \mathbf{x} \partial \dot{\mathbf{x}}}$ (derivatives w.r.t. x and x_dot of lagrangian) to be a bit ambiguous. [Since it's at least the case for scalars](https://en.wikipedia.org/wiki/Symmetry_of_second_derivatives), my (misplaced) gut instinct was that the order of derivatives shouldn't matter, in which case it's fine. But after looking into it, this is not the case. 

Now, to be clear, this is a skill issue :), but in:
 https://jax.readthedocs.io/en/latest/jax-101/04-advanced-autodiff.html#higher-order-derivatives 
chaininig of grad() calls is shown for the single-variable case, and how the Jacobian of the gradient computes the Hessian is also shown and verified. Below the Hessian case is where one *could* add a similar example, but with the Jacobian being taking w.r.t. a different vector than the gradient was taken against (and show how the ordering matters). Since one is the transpose of the other in something that, at least to me, at first glance seems like it should commute, it feels like an easy pitfall.

That's pretty much my 2 cents, but I understand that it might be too specific. Anyway, thanks for the reply!
